src/shims/x86/sse42.rs

```diff
@@ -0,0 +1,454 @@
+            // These functions return `1` if the string end has been reached and `0` otherwise.
+            // Since these functions define the string length implicitly, it is equal to a
+            // search for a null-terminator (see `deconstruct_args` for more details).
+            // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=924,925
+            "pcmpistriz128" | "pcmpistris128" => {
+                let [str1, str2, imm] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let imm = this.read_scalar(imm)?.to_u8()?;
+
+                let str = if unprefixed_name == "pcmpistris128" { str1 } else { str2 };
+                let array_layout = if imm & USE_WORDS != 0 {
+                    this.layout_of(Ty::new_array(this.tcx.tcx, this.tcx.types.u16, 8))?
+                } else {
+                    this.layout_of(Ty::new_array(this.tcx.tcx, this.tcx.types.u8, 16))?
+                };
+                let str = str.transmute(array_layout, this)?;
+                let result = implicit_len(this, &str, imm)?.is_some();
+
+                this.write_scalar(Scalar::from_i32(i32::from(result)), dest)?;
+            }
+
+            // Used to implement the `_mm_cmpestrz` and the `_mm_cmpestrs` functions.
+            // These functions return 1 if the explicitly passed string length is smaller
+            // than 16 for byte-sized operands or 8 for word-sized operands.
+            // https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#ig_expand=1046,1047
+            "pcmpestriz128" | "pcmpestris128" => {
+                let [_, len1, _, len2, imm] =
+                    this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
+                let len = if unprefixed_name == "pcmpestris128" { len1 } else { len2 };
+                let len = this.read_scalar(len)?.to_i32()?;
+                let imm = this.read_scalar(imm)?.to_u8()?;
+                this.write_scalar(
+                    Scalar::from_i32(i32::from(len < default_len::<i32>(imm))),
+                    dest,
+                )?;
+            }
+            _ => return Ok(EmulateItemResult::NotSupported),
+        }
+        Ok(EmulateItemResult::NeedsReturn)
+    }
+}
+
+/// The main worker for the string comparison intrinsics, where the given
+/// strings are analyzed according to the given immediate byte.
+///
+/// # Arguments
+///
+/// * `this` - The context under which interpretation is happening.
+/// * `str1` - The first string argument. It is always a length 16 array of bytes
+///   or a length 8 array of words.
+/// * `str2` - The second string argument. It is always a length 16 array of bytes
+///   or a length 8 array of words.
+/// * `len` is the length values of the supplied strings.
+///   If the value is `None`, the length of a string is determined by the first
+///   null value inside the string.
```

Comment:
I thought the strings have length 16...? That's what the docs for `str1`/`str2` seem to say.

Reply:
There are two length types at play here. One the one hand, there is the SIMD length (16 or 8) and the string length (how much of the input value is used). I've clarified the comment.

Reply:
Okay so these intrinsics work for strings of length "up to 16"? (or 8, for the two-byte mode)

Reply:
In word mode, I assume `len` is in units of words, not bytes? That would be good to clarify.

Reply:
Yes. For example, if you want to use use the substring functionality and search for the word "Rust", you'll need a way to specify that only the first four characters inside the argument are relevant and the remaining 12/4 are not.